operator/buffer/disk.go

```diff
@@ -42,26 +49,348 @@ func NewDiskBufferConfig() *DiskBufferConfig {
+	writeReady *sync.Cond
+	// closed is a bool indicating if the buffer is closed
+	closed bool
+
+	maxChunkDelay time.Duration
+	maxChunkSize  uint
 }
 
+// entryBufInitialSize is the initial size of the internal buffer that an entry
+// is written to
+const entryBufInitialSize = 1 << 10
+
 // Add adds an entry onto the buffer.
-// Is a blocking call if the buffer is full
-func (m *DiskBuffer) Add(ctx context.Context, e *entry.Entry) error {
+// Will block if the buffer is full
+func (d *DiskBuffer) Add(ctx context.Context, e *entry.Entry) error {
+	d.mux.Lock()
+	defer d.mux.Unlock()
+
+	if d.closed {
+		return ErrBufferClosed
+	}
+
+	bufBytes := make([]byte, 0, entryBufInitialSize)
+	bufBytes, err := marshalEntry(bufBytes, e)
+	if err != nil {
+		return err
+	}
+
+	if len(bufBytes) > int(d.maxSize) {
+		return ErrEntryTooLarge
+	}
+
+	for !d.canFitInBuffer(len(bufBytes)) {
+		err := d.compact()
+		if err != nil {
+			return err
+		}
+
+		if d.canFitInBuffer(len(bufBytes)) {
+			break
+		}
+
+		// Wait for a reader to tell us we can (maybe) write!
+		waitCondWithCtx(ctx, d.writeReady)
+
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("got context done: %w", ctx.Err())
+		default:
+		}
+
+		if d.closed {
+			return ErrBufferClosed
+		}
+	}
+
+	curEnd, err := d.f.Seek(0, io.SeekEnd)
+	if err != nil {
+		return err
+	}
+
+	// Assert for sanity; This should never occur
+	if curEnd != int64(d.end) {
+		return fmt.Errorf("the current EOF (%d) was not equal to the expected EOF (%d)", curEnd, d.end)
+	}
+
+	_, err = d.f.Write(bufBytes)
+	if err != nil {
+		return err
+	}
+
+	d.end += int64(len(bufBytes))
+
+	// Signal to a potentially waiting reader that there is an entry to read
+	d.readReady.Signal()
+
 	return nil
 }
 
 // Read reads from the buffer.
-// Read will block until the there are MaxChunkSize entries or we have block as long as MachChunkDelay.
-func (m *DiskBuffer) Read(ctx context.Context) ([]*entry.Entry, error) {
-	return nil, nil
+// Read will block until the there are maxChunkSize entries or the duration maxChunkDelay has passed.
+func (d *DiskBuffer) Read(ctx context.Context) ([]*entry.Entry, error) {
+	entries := make([]*entry.Entry, 0)
+	timer := time.NewTimer(d.maxChunkDelay)
+	defer timer.Stop()
+
+	d.mux.Lock()
+	defer d.mux.Unlock()
+
+	if d.closed {
+		return nil, ErrBufferClosed
+	}
+
+	for len(entries) < int(d.maxChunkSize) {
+		for d.end <= int64(d.metadata.StartOffset) {
+			// No entries to read
+			timerDone := waitCondWithCtxAndTimer(ctx, d.readReady, timer)
+			if timerDone {
+				return entries, nil
+			}
+
+			select {
+			case <-ctx.Done():
+				return nil, fmt.Errorf("got context done: %w", ctx.Err())
+			default:
+			}
+
+			if d.closed {
+				return nil, ErrBufferClosed
+			}
+		}
+
+		// Seek to the start position of the file
+		_, err := d.f.Seek(int64(d.metadata.StartOffset), io.SeekStart)
+		if err != nil {
+			return nil, err
+		}
+
+		var entry entry.Entry
+		dec := json.NewDecoder(d.f)
```

Comment:
Using JSON here; I looked into gob, but I don't think it's right for this purpose. It's self-describing, so it puts a whole header in every time it writes. It's better for long streams, but we don't really have that here (every entry is essentially standalone).